MINIFICPP-740: Add ability to run NiFi processors from Java and Python

phrocker opened the pull request:
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

phrocker commented:
not satisfied that I put all java code underneath one package. that was pure laziness. I will move those out at some point ( either in a follow on ticket or in this one )

phrocker commented on CMakeLists.txt:
```diff
@@ -487,6 +525,10 @@ if (ENABLE_BUSTACHE)
 endif()
 
+if (ENABLE_PYTHON)
+     #createExtension(PYTHON-EXTENSIONS "PYTHON EXTENSIONS" "This enables python processor support." "extensions/pyprocess" "${TEST_DIR}/pyproc-tests")
```
ah this was left in.

apiri commented:
reviewing

phrocker commented:
> reviewing

good luck

apiri commented:
This is going to be the best functionality we add in 2020!

phrocker commented on extensions/jni/ExecuteJavaProcessor.h:
```diff
@@ -0,0 +1,301 @@
+
+    JniSessionFactory *factory = new JniSessionFactory(ptr, java_servicer_, obj);
+
+    session_factories_.push_back(factory);
```
This is something I'm still testing. may not need more than one, in which case I Just applied a setter -- but will lock if this persists.

phrocker replied:
I added the lock in the event I forget to remove this code -- but it seems like I can get away with just one ( hence why it was a setter versus being called add ). Please don't resolve this convo so it's a placesetter so I can respond in the even the code gets removed. Thanks.

phrocker commented on extensions/jni/ExecuteJavaProcessor.cpp:
```diff
@@ -0,0 +1,221 @@
+  init = java_servicer_->loadClass("org/apache/nifi/processor/JniInitializationContext");
+
+  if (context_instance_ != nullptr) {
+    java_servicer_->attach()->DeleteGlobalRef(context_instance_);
```
It cannot. The lifetime of the object is inherently tied to a context, session, and flow file within Java.

phrocker replied:
Oh and to give more context: ProcessContext is probably the only object where we might be able to use RAII. The other objects ( flow file, session, and session factory we cannot ). As a result I made the painful decision to return the raw jobject versus a wrapped object. There are some things I'd like to improve in this class but hope to do so in subsequent releases.

phrocker commented on extensions/jni/ExecuteJavaProcessor.cpp:
```diff
@@ -0,0 +1,221 @@
+
+  jni_logger_ref_.logger_reference_ = nifi_logger_;
+
+  std::string dir_name_;
```
Thanks!

phrocker commented on extensions/jni/JVMCreator.cpp:
```diff
@@ -0,0 +1,129 @@
+        }
+        if (S_ISDIR(statbuf.st_mode)) {
+          // if this is a directory
+          if (strcmp(d_name.c_str(), "..") != 0 && strcmp(d_name.c_str(), ".") != 0) {
```
That's a good point. I copied this from elsewhere in the code as it did something similar, but could have easily used stringutils for that.

phrocker commented on extensions/jni/jvm/JVMLoader.h:
```diff
@@ -0,0 +1,514 @@
+    }
+
+    std::string modifiedName = name;
+    modifiedName = utils::StringUtils::replaceAll(modifiedName, "/", ".");
```
Not necessarily ( and not always ) but I will add a comment here. Thanks

phrocker commented on extensions/jni/JNILoader.h:
```diff
@@ -0,0 +1,77 @@
+   * @return class name for the processor.
+   */
+  virtual std::vector<std::string> getClassNames() {
+    std::vector<std::string> class_names;
```
Sure, that won't have any negatives at all.

phrocker commented on extensions/jni/jvm/JVMLoader.h:
```diff
@@ -0,0 +1,514 @@
+      std::vector<std::string> options;
+      for (const auto &path : pathVector) {
+        if (str.str().length() > 0) {
+#ifdef WIN32
```
Yep, just trying to limit trailing characters.

arpadboda replied:
Does pathVector contain empty strs?

phrocker commented on extensions/jni/jvm/JniFlowFile.cpp:
```diff
@@ -0,0 +1,171 @@
+
+  auto ff = ptr->get();
+  if (ff == nullptr) {
+    minifi::jni::ThrowJava(env, "Calling function on null flow file");
```
I'm going through and doing cleanup now I turned extracted a lot of those string references to macros -- but I just saw this in an E-mail ( haven't gone through the rest )  -- just wanted to comment that I really like this idea. Thanks!

arpadboda commented on extensions/script/README.md:
````diff
@@ -0,0 +1,64 @@
+	  processor.setSupportsDynamicProperties()
+	  processor.addProperty("property name","description","default value", True, False)
+```
+
````
I made a previous comment here, which I don't see now, so sorry for the duplication in case that happens!

Things I saw in the code and worth documenting in my opinion:
-getProperty call supports getting value of dynamic properties, too.  (as it's different in C++, I think it would be beneficial to have it documented)
-Python processors support success and failure relationships, these are the ones the script can send to. 

A question: what happens in case there is a Python exception thrown in "onTrigger"?
I wonder if a rollback should be triggered.

phrocker replied:
It will yield, as execute script does, if the python writer doesn't actually do the legwork of rolling back.

arpadboda commented on extensions/script/python/PythonCreator.cpp:
```diff
@@ -0,0 +1,129 @@
+        }
+        if (S_ISDIR(statbuf.st_mode)) {
+          // if this is a directory
+          if (strcmp(d_name.c_str(), "..") != 0 && strcmp(d_name.c_str(), ".") != 0) {
```
The same strcmp usage here as in Java

arpadboda commented on libminifi/include/utils/file/FileManager.h:
```diff
@@ -79,9 +95,15 @@ class FileManager {
+    std::string file_name = tmpDir + FILE_SEPARATOR + non_repeating_string_generator_.generate();
     while (!verify_not_exist(file_name)) {
-      file_name = "/tmp/" + non_repeating_string_generator_.generate();
+      file_name = tmpDir + FILE_SEPARATOR + non_repeating_string_generator_.generate();
```
Filesystem in cpp17 is really missed here... :(

arpadboda commented on thirdparty/pybind11/include/pybind11/pytypes.h:
```diff
@@ -292,7 +292,7 @@ class error_already_set : public std::runtime_error {
     /// Python error indicator will be cleared.
     error_already_set() : std::runtime_error(detail::error_string()) {
-        PyErr_Fetch(&type.ptr(), &value.ptr(), &trace.ptr());
+        //PyErr_Fetch(&type.ptr(), &value.ptr(), &trace.ptr());
```
Don't we need this?

phrocker replied:
I created a follow on ticket to explore this further -- but this was causing problems. I read several old tickets where people removed this line in regards to triggering seg faults. Usually this is an issue elsewhere, but ExecuteScript was causing problems. In most cases it was fine because we weren't really handling python exceptions, but I created a follow on because at some point I would like to better deal with that.

phrocker commented on extensions/jni/jvm/JniReferenceObjects.h:
```diff
@@ -0,0 +1,374 @@
+    return ff_object;
+  }
+
+  bool operator==(const JniFlowFile &other) {
```
Ugh...Eclipse doesn't auto add the const when it generates these functions. Thanks!

phrocker commented on libminifi/include/FlowController.h:
```diff
@@ -339,6 +339,8 @@ class FlowController : public core::controller::ControllerServiceProvider, publi
   // function to load the flow file repo.
   void loadFlowRepo();
 
+  void initializeJNI();
```
If a referenced processor isn't there we'll error out and tell them. Otherwise this will be a non issue.

phrocker commented on libminifi/include/core/VariableRegistry.h:
```diff
@@ -0,0 +1,105 @@
+        options.erase(std::remove(options.begin(), options.end(), opt), options.end());
+    }
+
+    // even if a white list is configured, remove the black listed fields
```
The comment does this a disservice but white listed fields allow selection only, blacklisted fields allow removal of fields from registry. 2 options, either it's an error condition to have both or we just remove blacklisted fields i the white list is specified. that was my preference for COOP.

phrocker commented on extensions/expression-language/Expression.cpp:
```diff
@@ -58,20 +58,27 @@ Expression make_static(std::string val) {
+      if ( registry && registry->getConfigurationProperty( attribute_id , result) ) {
+        return Value(result);
+      }
+      else {
```
Thanks!

phrocker commented on extensions/jni/ExecuteJavaProcessor.h:
```diff
@@ -0,0 +1,301 @@
+    if (registered_classes_.find(className) == std::end(registered_classes_)) {
+      auto cls = servicer->loadClass(className);
+      cls.registerMethods(env, signatures);
+      registered_classes_.insert(className);
```
added a comment as to why find was chosen

phrocker commented on extensions/jni/jvm/JavaControllerService.cpp:
```diff
@@ -0,0 +1,198 @@
+        }
+        if (S_ISDIR(statbuf.st_mode)) {
+          // if this is a directory
+          if (strcmp(d_name.c_str(), "..") != 0 && strcmp(d_name.c_str(), ".") != 0) {
```
Yep! Thanks!

phrocker commented on extensions/jni/jvm/JniLogger.cpp:
```diff
@@ -0,0 +1,106 @@
+  return true;
+}
+jboolean Java_org_apache_nifi_processor_JniLogger_isInfoEnabled(JNIEnv *env, jobject obj) {
+  minifi::jni::JniLogger *logger_ref = minifi::jni::JVMLoader::getPtr<minifi::jni::JniLogger>(env, obj);
```
Sorry I did some cleanup after I submitted review. I didn't realize you submitted this. I think these are addressed.

phrocker commented on extensions/jni/jvm/JniLogger.cpp:
```diff
@@ -0,0 +1,106 @@
+  return true;
+}
+
+void Java_org_apache_nifi_processor_JniLogger_warn(JNIEnv *env, jobject obj, jstring msg) {
```
Thanks for your suggestion.

phrocker commented on extensions/jni/jvm/JniReferenceObjects.h:
```diff
@@ -0,0 +1,374 @@
+
+  virtual void remove() override;
+
+  std::shared_ptr<core::FlowFile> get() {
```
I made some changes before seeing some of these. I think many of these were caught. But I may have missed something.

phrocker commented on extensions/jni/jvm/JniReferenceObjects.h:
```diff
@@ -0,0 +1,374 @@
+    return ref_ == other.ref_;
+  }
+
+  bool empty() {
```
I made some changes before seeing some of these. I think many of these were caught, including this one

phrocker commented on extensions/jni/jvm/JniReferenceObjects.h:
```diff
@@ -0,0 +1,374 @@
+
+ protected:
+  bool removed_;
+  std::mutex session_mutex_;
```
I made some changes before seeing some of these. I think many of these were caught, including this one

phrocker commented on extensions/jni/jvm/NarClassLoader.h:
```diff
@@ -0,0 +1,443 @@
+    }
+    {
+
+      jmethodID mthd = env->GetMethodID(class_ref_.getReference(), "getSignature", "(Ljava/lang/String;Ljava/lang/String;)Ljava/lang/String;");
```
Most likely. I'll catch it on another round of cleanup. thanks!

phrocker commented:
@apiri and @arpadboda I created a ticket for the travis test failure. It's transient and doesn't always occur https://issues.apache.org/jira/browse/MINIFICPP-743

arpadboda commented:
Looks good as far as I can tell! The comments I considered important are resolved.

@phrocker : could you make an exemption and add cleanups or anything else added as a separate commit instead of amend? 
The PR is very big, so it's difficult to find the changes. Thanks in advance!

phrocker commented:
@arpadboda sure thing. I'm doing self review and cleanup, so I'll have one more force push but after that it's all yours.

phrocker commented on extensions/jni/jvm/JniMethod.h:
```diff
@@ -0,0 +1,131 @@
+      method_ptr_ = std::unique_ptr<JNINativeMethod>(new JNINativeMethod[methods_.size()]);
+      size_ = methods_.size();
+      int i = 0;
+      for (auto &mthd : methods_) {
```
ha I don't know why I did that. Thanks for finding it!

phrocker commented on extensions/jni/nifi-framework-jni/src/main/java/org/apache/nifi/nar/JniUnpacker.java:
```diff
@@ -0,0 +1,398 @@
+
+            final File bundledDependencies = new File(unpackedNar, "NAR-INF/bundled-dependencies");
+
+            unpackBundleDocs(docsDirectory, mapping, bundleCoordinate, bundledDependencies);
```
The documentation provides useful info in the jar entries ( meta-inf/services/*) that could help us supplement service building the documentation for the agent manifest.

phrocker replied:
nm i don't think i could really get all the information I need without loading processors anyway. good point re the docs.

phrocker commented on extensions/jni/nifi-framework-jni/src/main/java/org/apache/nifi/nar/JniUnpacker.java:
```diff
@@ -0,0 +1,398 @@
+/*
```
oh man missed this. thanks

arpadboda commented:
@phrocker:
I don't like the fact that JNI API doesn't provide RAII wrapper from UTF string handling, so I quickly sratched one:
```
class JNIUTFString {
 public:
  JNIUTFString(JNIEnv * env, jstring jstr) {
    _env = env;
    _jstr = jstr;
    _cstr = _env->GetStringUTFChars(_jstr, NULL);
  }
  ~JNIUTFString() {
    _env->ReleaseStringUTFChars(_jstr, _cstr);
  }
  
  operator std::string() const { return _cstr; }
  
  const char * c_str() const { return _cstr; }
  
  friend std::ostream& operator<<(std::ostream& os, const JNIUTFString& jutfstr);
  
 private:
  JNIEnv * _env;
  const char * _cstr;
  jstring _jstr;
};

std::ostream& operator<<(std::ostream& os, const JNIUTFString& jutfstr)
{
    os << jutfstr.c_str();
    return os;
}
```

This can be used as an std::string, but you don't have to deal with conversion, releasing, etc, usage is simple:

```
  JNIUTFString jutf(env, jstr);
  std::set<std::string> test_set;
  test_set.insert(jutf);
  std::cout << jutf <<std::endl;
  std::cout << jutf.c_str() << std::endl;
```

Feel free to add if you like it.

phrocker commented:
@apiri I didn't resolve convos because I believe that is up to the author of that comment; however, I believe I addressed most issues. please take another gander. I'm not sure if that last comment was directed at me or Arpad, though.

apiri commented:
@phrocker sounds fair.  will do one last scan/build/test a bit later and we can move on.  thanks!
